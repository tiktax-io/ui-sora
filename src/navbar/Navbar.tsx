import React, { ForwardedRef, forwardRef, memo, useEffect, useState } from 'react'
import { useStyles } from './Navbar.styles'
import { HamburgerMenuProps, NavbarProps } from './Navbar.types'

/******************************************************************************
 * Hamburger menu icon. Displayed as 3 horizontal lines used as navigation.
 * @param {Color} color string of hex or rgba color to apply.
 * @param {() => void} onClick function to trigger onclick.
 * @returns Icon with the 3 horizontal lines representing a clickable menu.
 *****************************************************************************/
const HamburgerMenu = ({
  color = '#686868',
  onClick
}: HamburgerMenuProps) => (
  <span onClick={onClick} style={{ cursor: 'pointer', width: '30px' }}>
    <svg width="30" height="26" viewBox="0 0 30 26" fill="none" xmlns="http://www.w3.org/2000/svg">
      <title>HamburgerMenu</title>
      <path d="M0 2.14286C0 0.959597 0.959598 -9.53674e-07 2.14286 -9.53674e-07H27.8571C29.0424 -9.53674e-07 30 0.959597 30 2.14286C30 3.32812 29.0424 4.28571 27.8571 4.28571H2.14286C0.959598 4.28571 0 3.32812 0 2.14286ZM0 12.8571C0 11.6719 0.959598 10.7143 2.14286 10.7143H27.8571C29.0424 10.7143 30 11.6719 30 12.8571C30 14.0424 29.0424 15 27.8571 15H2.14286C0.959598 15 0 14.0424 0 12.8571ZM27.8571 25.7143H2.14286C0.959598 25.7143 0 24.7567 0 23.5714C0 22.3862 0.959598 21.4286 2.14286 21.4286H27.8571C29.0424 21.4286 30 22.3862 30 23.5714C30 24.7567 29.0424 25.7143 27.8571 25.7143Z" fill={color} />
    </svg>
  </span>
)

/******************************************************************************
 * Component to display on top, usually displaying logo and navigation options.
 * @param {ReactNode} children elements to render inside Navbar container.
 * @param {Color} color text and content color.
 * @param {string} dataTestId string to identify by id on tests. "data-testid".
 * @param {boolean} hasHamburgerMenu boolean, true if renders hamburger menu
 * icon to show/hide sidebar navigation.
 * @param {() => void | undefined} onHamburgerMenuClick function to trigger on
 * clicking hamburger menu icon.
 * @returns {ReactElement} HTML Header element wrapping content.
 * @see https://www.figma.com/file/GQCJ1nVFWevddz10mxmcgP/ui-sora?node-id=749%3A712
 *****************************************************************************/
const Navbar = memo(forwardRef(function({
  children,
  color,
  dataTestId,
  hasHamburgerMenu,
  onHamburgerMenuClick
}: NavbarProps, ref: ForwardedRef<HTMLElement> | undefined) {
  const { innerWidth } = window
  const [width, setWidth] = useState(innerWidth)

  useEffect(() => {
    window.addEventListener('resize', () => setWidth(window.innerWidth))
  }, [])

  const classes = useStyles()

  const ifHasHamburgerMenu = (hasHamburgerMenu !== false) && width < 1280 ||
    hasHamburgerMenu === true

  const hamburgerIconProps = {
    color: color,
    onClick: onHamburgerMenuClick
  }

  return (
    <header className={classes.container} data-testid={dataTestId} ref={ref}>
      <div className={classes.content}>
        {ifHasHamburgerMenu && <HamburgerMenu {...hamburgerIconProps} />}
        {children}
      </div>
    </header>
  )
}))

export default Navbar
