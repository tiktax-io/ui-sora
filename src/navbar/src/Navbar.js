'use strict'
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k
        var desc = Object.getOwnPropertyDescriptor(m, k)
        if (!desc || ('get' in desc ? !m.__esModule : desc.writable || desc.configurable)) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k]
            },
          }
        }
        Object.defineProperty(o, k2, desc)
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k
        o[k2] = m[k]
      })
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, 'default', { enumerable: true, value: v })
      }
    : function (o, v) {
        o['default'] = v
      })
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod
    var result = {}
    if (mod != null)
      for (var k in mod)
        if (k !== 'default' && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k)
    __setModuleDefault(result, mod)
    return result
  }
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
/* eslint-disable react/prop-types */
const react_1 = __importStar(require('react'))
const react_jss_1 = require('react-jss')
const Navbar_styles_1 = __importDefault(require('./Navbar.styles'))
const buttons_1 = require('../../buttons')
const icons_1 = require('../../icons')
/******************************************************************************
 * A layout wrapper in the form of a Dashboard. A left fixed div where
 * navigation elements can be rendered. On mobile hamburger menu will render
 * the side div. On narrow screens the side div has the enough width to render
 * icons instead of full nav tabs.
 * @param {unknown} props
 * @param {ref} ref
 * @returns {ReactElement} Functional component, returns elements to render.
 *****************************************************************************/
const Navbar = (0, react_1.memo)(
  (0, react_1.forwardRef)((props, ref) => {
    // const { ...restProps } = props
    const useStyles = (0, react_jss_1.createUseStyles)((0, Navbar_styles_1.default)(props))
    const classes = useStyles()
    const iconButtonProps = {
      appearance: 'minimal',
      border: 'none',
      'data-testid': 'navbar-hamburger-menu',
      icon: icons_1.MenuIcon,
      height: 40,
      onClick: () => (props.handleShowSidebar ? props.handleShowSidebar() : null),
    }
    return react_1.default.createElement(
      'div',
      { className: classes.navbar, ref: ref },
      react_1.default.createElement(
        'div',
        { className: classes.navbarContainer },
        !props.hideToggle && react_1.default.createElement(buttons_1.IconButton, Object.assign({}, iconButtonProps))
      )
    )
  })
)
exports.default = Navbar
