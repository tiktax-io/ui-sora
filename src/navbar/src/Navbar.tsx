/* eslint-disable react/prop-types */
import React, { memo, forwardRef, Ref } from 'react'
import { createUseStyles } from 'react-jss'
import styles from './Navbar.styles'
import { IconButton } from '../../buttons'
import { MenuIcon } from '../../icons'
import { IconButtonProps, NavbarProps } from '../../_types'

/******************************************************************************
 * A layout wrapper in the form of a Dashboard. A left fixed div where
 * navigation elements can be rendered. On mobile hamburger menu will render
 * the side div. On narrow screens the side div has the enough width to render
 * icons instead of full nav tabs.
 * @param {unknown} props
 * @param {ref} ref
 * @returns {ReactElement} Functional component, returns elements to render.
 *****************************************************************************/
const Navbar = memo(
  forwardRef((props: NavbarProps, ref: Ref<HTMLDivElement>) => {
    // const { ...restProps } = props
    const useStyles = createUseStyles(styles(props))
    const classes = useStyles()
    const iconButtonProps: IconButtonProps = {
      appearance: 'minimal',
      border: 'none',
      'data-testid': 'navbar-hamburger-menu',
      icon: MenuIcon,
      height: 40,
      onClick: () => (props.handleShowSidebar ? props.handleShowSidebar() : null),
    }

    return (
      <div className={classes.navbar} ref={ref}>
        <div className={classes.navbarContainer}>{!props.hideToggle && <IconButton {...iconButtonProps} />}</div>
      </div>
    )
  })
)

export default Navbar
