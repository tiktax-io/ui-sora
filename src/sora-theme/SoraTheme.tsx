import React, { memo } from 'react'
import { ThemeProvider } from 'react-jss'
import { SoraThemeProps } from './SoraTheme.types'

/******************************************************************************
 * Wraps children elements and styles them with the theme passed as prop.
 * @param {ReactNode} children JSX elements to be styled with context.
 * @param {Theme} theme object containing the styles.
 * @returns {React.FC} Wrapper with theme containing styles as a Context.
 *****************************************************************************/
const SoraTheme = memo(function SoraTheme({
  children,
  theme
}: SoraThemeProps) {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
})

export default SoraTheme
