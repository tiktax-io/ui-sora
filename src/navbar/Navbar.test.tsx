import React, { ReactNode } from 'react'
import ReactDOM from 'react-dom/client'
import { act } from 'react-dom/test-utils'
import { fireEvent, render, screen } from '@testing-library/react'
import Navbar from './Navbar'
import SoraTheme from '../sora-theme/SoraTheme'
import darkDefault from '../_styles/official_themes/dark_default'

describe('Jest can resize window width', () => {
  const defaultWidth = window.innerWidth

  it('Viewport width can be modified by Jest', () => {
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: 390
    })
    fireEvent.resize
    expect(window.innerWidth).toBe(390)
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: defaultWidth
    })
    fireEvent.resize
  })
  it('Viewport width is reseted after each test', () => {
    expect(window.innerWidth).not.toBe(390)
  })
})

describe('Navbar component', () => {
  const container = document.createElement('div')
  beforeEach(() => {
    document.body.appendChild(container)
  })
  /******************************************************************************
   * Reset DOM after each test
   *****************************************************************************/
   afterEach(() => {
    document.body.removeChild(container)
  })

  test('Renders multiple children without crashing', () => {
    render(
      <Navbar>
        <div>one</div>
        <div>two</div>
      </Navbar>
    )
    const oneElement = screen.getByText(/one/i)
    const twoElement = screen.getByText(/two/i)
    expect(oneElement).toBeInTheDocument()
    expect(twoElement).toBeInTheDocument()
  })

  test('HamburgerMenuIcon is shown by default on small screens', () => {
    const defaultWidth = window.innerWidth
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 390
      })
      fireEvent.resize
    })
    render(
      <Navbar>
        <div>one</div>
        <div>two</div>
      </Navbar>
    )
    const icon = screen.getByTitle('HamburgerMenu')
    expect(icon).toBeInTheDocument()
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: defaultWidth
    })
    fireEvent.resize
  })

  test('HamburgerMenuIcon is hidden on desktop screens by default', () => {
    const defaultWidth = window.innerWidth
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1500
      })
      fireEvent.resize
    })

    render(<Navbar/>)
    expect(screen.queryByTitle('HamburgerMenu')).toBeNull()
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: defaultWidth
      })
      fireEvent.resize
    })
  })

  test('Accepts an `onHamburgerMenuClick` handler that gets called upon clicking on icon', () => {
    const defaultWidth = window.innerWidth
    const handleClick = jest.fn()

    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 390
      })
      fireEvent.resize
    })

    render(<Navbar onHamburgerMenuClick={handleClick}/>)
    expect(handleClick).toHaveBeenCalledTimes(0)
    fireEvent.click(screen.getByTitle('HamburgerMenu'))
    expect(handleClick).toHaveBeenCalledTimes(1)
    fireEvent.click(screen.getByTitle('HamburgerMenu'))
    expect(handleClick).toHaveBeenCalledTimes(2)

    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: defaultWidth
      })
      fireEvent.resize
    })
  })

  test('HamburgerMenuIcon is hidden on small screens if specified by prop', () => {
    const defaultWidth = window.innerWidth
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 390
      })
      fireEvent.resize
    })
    render(
      <Navbar hasHamburgerMenu={false} />
    )
    const icon = screen.queryByTitle('HamburgerMenu')
    expect(icon).toBeNull()
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: defaultWidth
    })
    fireEvent.resize
  })

  test('HamburgerMenuIcon is shown on desktop screens if specified by prop', () => {
    const defaultWidth = window.innerWidth
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1500
      })
      fireEvent.resize
    })
    render(
      <Navbar hasHamburgerMenu />
    )
    const icon = screen.getByTitle('HamburgerMenu')
    expect(icon).toBeInTheDocument()
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: defaultWidth
    })
    fireEvent.resize
  })

  test('Navbar supports theme customisation', () => {
    render(
      <div style={{ alignItems: 'center', display: 'flex', gap: '20px' }}>
        <SoraTheme theme={darkDefault}>
          <Navbar dataTestId='Customized1' />
          <Navbar dataTestId='Customized2' />
        </SoraTheme>
        <Navbar dataTestId='Default1' />
        <Navbar dataTestId='Default2' />
      </div>
    )
    const CustomNavbar = screen.getByTestId(/Customized1/i).getAttribute('class')
    const CustomNavbar2 = screen.getByTestId(/Customized2/i).getAttribute('class')
    const Default1 = screen.getByTestId(/Default1/i).getAttribute('class')
    const Default2 = screen.getByTestId(/Default2/i).getAttribute('class')
    expect(CustomNavbar).toMatch(`${CustomNavbar2}`)
    expect(Default1).toMatch(`${Default2}`)
    expect(CustomNavbar).not.toMatch(`${Default1}`)
    expect(CustomNavbar).not.toMatch(`${Default2}`)
  })

})
