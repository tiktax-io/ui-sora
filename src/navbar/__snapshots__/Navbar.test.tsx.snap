// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of styles are matched 1`] = `
<header
  className="container-0-2-1"
  data-testid="test-snapshot"
>
  <div
    className="content-0-2-2"
  >
    <span
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
          "width": "30px",
        }
      }
    >
      <svg
        fill="none"
        height="26"
        viewBox="0 0 30 26"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          HamburgerMenu
        </title>
        <path
          d="M0 2.14286C0 0.959597 0.959598 -9.53674e-07 2.14286 -9.53674e-07H27.8571C29.0424 -9.53674e-07 30 0.959597 30 2.14286C30 3.32812 29.0424 4.28571 27.8571 4.28571H2.14286C0.959598 4.28571 0 3.32812 0 2.14286ZM0 12.8571C0 11.6719 0.959598 10.7143 2.14286 10.7143H27.8571C29.0424 10.7143 30 11.6719 30 12.8571C30 14.0424 29.0424 15 27.8571 15H2.14286C0.959598 15 0 14.0424 0 12.8571ZM27.8571 25.7143H2.14286C0.959598 25.7143 0 24.7567 0 23.5714C0 22.3862 0.959598 21.4286 2.14286 21.4286H27.8571C29.0424 21.4286 30 22.3862 30 23.5714C30 24.7567 29.0424 25.7143 27.8571 25.7143Z"
          fill="#0070ac"
        />
      </svg>
    </span>
    <span>
      Children
    </span>
  </div>
</header>
`;
